sqllogictest/src/parser.rs

```diff
@@ -603,11 +614,87 @@ fn parse_file_inner(loc: Location) -> Result<Vec<Record>, ParseError> {
 
 #[cfg(test)]
 mod tests {
-    use crate::parse_file;
+    use difference::{Changeset, Difference};
+
+    use super::*;
 
     #[test]
     fn test_include_glob() {
         let records = parse_file("../examples/include/include_1.slt").unwrap();
-        assert_eq!(14, records.len());
+        assert_eq!(16, records.len());
+    }
+
+    #[test]
+    fn test_basic() {
+        parse_roundtrip("../examples/basic/basic.slt")
+    }
+
+    /// Parses the specified file into Records, and ensures the
+    /// results of unparsing them are the same
+    ///
+    /// Prints a hopefully useful message on failure
+    fn parse_roundtrip(filename: impl AsRef<Path>) {
+        let filename = filename.as_ref();
+        let input_contents = std::fs::read_to_string(filename).expect("reading file");
+
+        let records = parse_file(filename).expect("parsing to complete");
+
+        println!("Parsed records:\n{:#?}", records);
+
+        let unparsed = records
+            .into_iter()
+            .map(|r| record_to_string(r))
+            .collect::<Vec<_>>();
+
+        let output_contents = unparsed.join("\n");
+
+        let changeset = Changeset::new(&input_contents, &output_contents, "\n");
+
+        assert!(
+            no_diffs(&changeset),
+            "Mismatch for {:?}\n\
+                 *********\n\
+                 diff:\n\
+                 *********\n\
+                 {}\n\n\
+                 *********\n\
+                 output:\n\
+                 *********\n\
+                 {}\n\n",
+            filename,
+            UsefulDiffDisplay(&changeset),
+            output_contents,
+        );
+    }
+
+    fn record_to_string(record: Record) -> String {
+        if matches!(&record, Record::Statement { .. } | Record::Query { .. }) {
+            // the statement parser includes a newline between the items but the display
+            // output does not, so add it here
+            // Not sure about this one
+            format!("{}\n", record)
```

Comment:
I found this to be necessary to recreate the original `.slt` file.

I think it would more naturally go in the `impl Display` itself, however I didn't want to mess up any of your existing tests. What do you think @xxchan ?

Here is what the output looks like without this clause:

<details>
  <summary>Click me</summary>
  
```
running 2 tests
test parser::tests::test_include_glob ... ok
test parser::tests::test_basic ... FAILED

failures:

---- parser::tests::test_basic stdout ----

*********
diff:
*********
statement ok
create table t(v1 int not null, v2 int not null, v3 int not null)
-   
statement ok
insert into t values(1,4,2), (2,3,3), (3,4,4), (4,3,5)

-   
query I
select * from example_basic
----
Alice
Bob
Eve
-   
statement ok
drop table t
-   
# The error message is: Hey you got FakeDBError!

statement error
give me an error
-   
statement error FakeDB
give me an error
-   
statement error (?i)fakedb
give me an error
-   
statement error Hey you
give me an error
-   
# statement is expected to fail with error: "Hey we", but got error: "Hey you got FakeDBError!"
# statement error Hey we
# give me an error

# query error is actually equivalent to statement error
query error Hey you
give me an error
-   
hash-threshold 3

query I
select * from example_basic
----
Alice
Bob
Eve
-   
hash-threshold 1

query I
select * from example_basic
----
3 values hashing to b5b44edac84d34d6af3be2a88bfae352
-   
hash-threshold 0

query I
select * from example_basic
----
Alice
Bob
Eve
-   


*********
output:
*********
statement ok
create table t(v1 int not null, v2 int not null, v3 int not null)
statement ok
insert into t values(1,4,2), (2,3,3), (3,4,4), (4,3,5)

query I
select * from example_basic
----
Alice
Bob
Eve
statement ok
drop table t
# The error message is: Hey you got FakeDBError!

statement error
give me an error
statement error FakeDB
give me an error
statement error (?i)fakedb
give me an error
statement error Hey you
give me an error
# statement is expected to fail with error: "Hey we", but got error: "Hey you got FakeDBError!"
# statement error Hey we
# give me an error

# query error is actually equivalent to statement error
query error Hey you
give me an error
hash-threshold 3

query I
select * from example_basic
----
Alice
Bob
Eve
hash-threshold 1

query I
select * from example_basic
----
3 values hashing to b5b44edac84d34d6af3be2a88bfae352
hash-threshold 0

query I
select * from example_basic
----
Alice
Bob
Eve
```

</details>

Reply:
BTW why it doesn't work to add newline to every record here in `record_to_string`?

Reply:
Oh it is because newline is consumed when parsing `Statement` and `Query` 🤔

Reply:
Now I think it may be reasonable to add trailing newline in `Display` for `Query` and `Statement`.. Since empty line is their terminator. Other records doesn't require empty lines in between.

Reply:
> Oh it is because newline is consumed when parsing Statement and Query 🤔

Yes, I think this is the reason. 

Logically a `Statement` and `Query` must *always* be followed by a blank line (that newline isn't the equivalent of `println`). Much like a multi line comment contains a "\n" within it even though there `impl Display`

Or put another way, the following would not be parsed correctly. 
```
query II
select a, b from foo
---
1 2
query II
select a, b from bar
---
3 4
```


`Query` must always end with a newline (other than the last one in the file):

```
query II
select a, b from foo
---
1 2

select a, b from bar
---
3 4

```

Reply:
> BTW why it doesn't work to add newline to every record here in record_to_string?

You if you do that, then there are extra newlines between every record (as I am already using `writeln`)